### DISCUSSION TITLE
Pencil Tracing

### DISCUSSION THREAD

**Discussion by evancasey** [visioncortex/vtracer#63] [General]
It's listed on the readme that "Pencil tracing" (i.e. centerline tracing) is on the roadmap. I'm working on a project where this could be super valuable and might be able to contribute to the development/discussion.

But where to even start? Aside from Autotrace and a few SIGGRAPH papers I haven't seen much work in this area. It would be great if someone with experience here could summarize the common approaches or provide some direction.

**Comment by tyt2y3** [comment]
I have had various prototypes I wrote many years ago lying around, with various degrees of success.
I think it heavily depends on the input you have. If they are more or less idealized, like electronic circuits, it is easier to exploit a few invariants there. But if you scanned a hand drawn drawing it would be very difficult and computationally expensive.

So what does your input look like?

**Comment by tyt2y3** [comment]
![son stroke matlab](https://github.com/visioncortex/vtracer/assets/1782664/bf5926a9-f5d4-4a78-a6bc-d1f06bc5aaea)
^ raw skeleton

![son stroke ours](https://github.com/visioncortex/vtracer/assets/1782664/fba12974-bbd2-41c2-bcc1-06d5bec8a13a)
^ extracted skeleton

one approach is to first skeletonize the shape and then prune the tree until you get the central axis

**Comment by evancasey** [comment]
Interesting, would you skeletonize before or after vectorization?

I'm interesting in applying this to hand-drawn animation/cartoon images. The line quality is usually very good (see attached), but there a cases where the drawing is done with a thicker pencil which could make it harder.

But even getting centerline tracing on an image like this would be huge...

![final-122](https://github.com/visioncortex/vtracer/assets/1900535/c8650242-8068-401c-b7b2-01274cdfbd07)

**Reply by evancasey** [reply]
I'm most worried about getting all the joints to line up, in which case the vectorizer might treat the line as disjoint paths

**Comment by tyt2y3** [comment]
![image](https://github.com/visioncortex/vtracer/assets/1782664/8f886a1f-d589-47ff-8510-65ccb2b63c30)

the skeleton does not have to be perfect to begin with, it just needs to tell us the topology roughly

the above "rough" algorithm was included in https://docs.rs/visioncortex/latest/visioncortex/struct.BinaryImage.html#method.to_skeleton

but it would fall short when the line is very thick:

![image](https://github.com/visioncortex/vtracer/assets/1782664/394f3d34-7d31-4d9b-bbd2-2388d1ac9505)

**Reply by tyt2y3** [reply]
> skeletonize before or after vectorization

after clustering, as raw polygons; before curve fitting

**Comment by tyt2y3** [comment]
I see you, the thickness of your lines are very manageable!

**Reply by evancasey** [reply]
nice!! I'll give skeletonizing a try and post the results here :)

**Reply by evancasey** [reply]
How thin does the line need to for vtracer to not output double lines? Or would I need to also change the vectorization code to merge double lines.

**Comment by tyt2y3** [comment]
One thing I did was to first detect all junction (or intersections) and then decompose the shapes into simplier patches before skeletonizing

![image](https://github.com/visioncortex/vtracer/assets/1782664/783ba261-6c1b-4a98-b4b1-99fa3e095229)
^ intersection points marked in red

![image](https://github.com/visioncortex/vtracer/assets/1782664/6bd75807-a998-4910-bb7d-e0f80a3b3eba)
^ gray: input; black: final skeleton

**Reply by tyt2y3** [reply]
so the complexity in your particular sample is more or less doable

**Reply by Tom1827** [reply]
> One thing I did was to first detect all junction (or intersections) and then decompose the shapes into simplier patches before skeletonizing
> 
> ![image](https://private-user-images.githubusercontent.com/1782664/283215358-783ba261-6c1b-4a98-b4b1-99fa3e095229.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTQ0MDY1NzEsIm5iZiI6MTc1NDQwNjI3MSwicGF0aCI6Ii8xNzgyNjY0LzI4MzIxNTM1OC03ODNiYTI2MS02YzFiLTRhOTgtYjRiMS05OWZhM2UwOTUyMjkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDgwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA4MDVUMTUwNDMxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OGNlZjQyNDI0OTZiOThmYWI4YmJhYjZlMWJmMDg4NDY0ZmZiMjEzZmY2ODM4MTI1OWViMTEzYzU5YTEzMmRiMCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.b0sGRy4CjBkTw_KbNNLJ5bIhzpjoGLanTNEJUbHgMsg) ^ intersection points marked in red
> 
> ![image](https://private-user-images.githubusercontent.com/1782664/283215758-6bd75807-a998-4910-bb7d-e0f80a3b3eba.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTQ0MDY1NzEsIm5iZiI6MTc1NDQwNjI3MSwicGF0aCI6Ii8xNzgyNjY0LzI4MzIxNTc1OC02YmQ3NTgwNy1hOTk4LTQ5MTAtYmI3ZC1lMGY4MGEzYjNlYmEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDgwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA4MDVUMTUwNDMxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OTExNjg2NmMzMTVjYmQxNjExMmNiZjJlYWIxYzYyZjQ3MTRiMjEwZDA4NWUzMTU5ZjczZjRmYmE3YjkyNTZkYiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.te5C9VRa3rol8rh9UHKvw3RcttWXDEYdYumxotrFmpM) ^ gray: input; black: final skeleton

I'm also super-interested in pencil tracing being available; would you be willing to share the code behind this example image for experimentation?

**Comment by evancasey** [comment]
Sorry, forgot to respond here. How are you using the intersections? Not sure I follow this:

> One thing I did was to first detect all junction (or intersections) and then decompose the shapes into simplier patches before skeletonizing

I do think making the problem simpler with junction detection is a fruitful approach, but it seems like your output here has gaps in the junctions

**Comment by evancasey** [comment]
As a general approach, why not start with the paths and reduce them down into pencil strokes, eg: https://www.youtube.com/watch?v=tEZF7Fo6Fpg

**Reply by tyt2y3** [reply]
Depending on the input, it won't work in the general case 
![image](https://github.com/visioncortex/vtracer/assets/1782664/7f8e2fd9-3d7a-4f2f-84e9-4d4001a90bc4)

**Comment by faces-of-eth** [comment]
Noob here. Would also love simple pencil tracing. Any branches being worked on a noob could hack on?

**Comment by tyt2y3** [comment]
I’ve been planning this for a while, but so far not starting yet.
________________________________

**Comment by 0xd1cf001e** [comment]
Here's a solution that uses a divide-and-conquer approach, though it operates on raw pixels rather than vectorized images, it produces very good result even for complex characters: https://github.com/LingDong-/skeleton-tracing

**Reply by tyt2y3** [reply]
this is an interesting algorithm. thank you for sharing. I'll see if I can use that!
